File: src/content/code-samples/authentication/full-auth-flow.java
package com.example.auth;

import com.scalekit.ScalekitClient;
import com.scalekit.models.AuthorizationUrlOptions;
import com.scalekit.models.AuthenticationResponse;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.web.bind.annotation.*;
import javax.servlet.http.HttpSession;

@SpringBootApplication
@RestController
public class AuthApplication {

    // Initialize Scalekit client
    private final ScalekitClient scalekitClient;

    public AuthApplication() {
        this.scalekitClient = new ScalekitClient(
            System.getenv("SCALEKIT_ENV_URL"),
            System.getenv("SCALEKIT_CLIENT_ID"),
            System.getenv("SCALEKIT_CLIENT_SECRET")
        );
    }

    public static void main(String[] args) {
        SpringApplication.run(AuthApplication.class, args);
    }

    // Step 1: Generate authorization URL
    @GetMapping("/auth/login")
    public String login(@RequestParam String organizationId) {
        AuthorizationUrlOptions options = new AuthorizationUrlOptions()
            .organizationId(organizationId)
            .state("random-state-string"); // Use a secure random string

        String authorizationUrl = scalekitClient.getAuthorizationUrl(
            "http://localhost:8080/auth/callback",
            options
        );

        return "redirect:" + authorizationUrl;
    }

    // Step 2: Handle OAuth callback
    @GetMapping("/auth/callback")
    public String callback(
        @RequestParam String code,
        @RequestParam String state,
        HttpSession session
    ) {
        try {
            // Exchange authorization code for tokens
            AuthenticationResponse result = scalekitClient.authenticateWithCode(
                code,
                "http://localhost:8080/auth/callback"
            );

            // Access user information
            var user = result.getUser();
            var idTokenClaims = result.getIdTokenClaims();

            // Step 3: Create session and store user info
            session.setAttribute("userId", user.getId());
            session.setAttribute("email", user.getEmail());
            session.setAttribute("organizationId", idTokenClaims.getOrgId());

            return "redirect:/dashboard";
        } catch (Exception error) {
            System.err.println("Authentication failed: " + error.getMessage());
            return "redirect:/login?error=auth_failed";
        }
    }

    // Step 4: Protect routes with authentication check
    @GetMapping("/dashboard")
    public String dashboard(HttpSession session) {
        String userId = (String) session.getAttribute("userId");
        if (userId == null) {
            return "redirect:/login";
        }

        String email = (String) session.getAttribute("email");
        return "Welcome, " + email + "!";
    }

    // Step 5: Handle logout
    @PostMapping("/auth/logout")
    public String logout(HttpSession session) {
        session.invalidate();
        return "redirect:/login";
    }
}

File: src/content/code-samples/authentication/full-auth-flow.py
from scalekit import ScalekitClient
from flask import Flask, request, redirect, session
import os

# Initialize Scalekit client
scalekit = ScalekitClient(
    env_url=os.getenv('SCALEKIT_ENV_URL'),
    client_id=os.getenv('SCALEKIT_CLIENT_ID'),
    client_secret=os.getenv('SCALEKIT_CLIENT_SECRET')
)

app = Flask(__name__)
app.secret_key = os.getenv('SESSION_SECRET')

# Step 1: Generate authorization URL
@app.route('/auth/login')
def login():
    organization_id = request.args.get('organization_id')

    authorization_url = scalekit.get_authorization_url(
        redirect_uri='http://localhost:5000/auth/callback',
        options={
            'organization_id': organization_id,
            'state': 'random-state-string'  # Use a secure random string
        }
    )

    return redirect(authorization_url)

# Step 2: Handle OAuth callback
@app.route('/auth/callback')
def callback():
    code = request.args.get('code')
    state = request.args.get('state')

    try:
        # Exchange authorization code for tokens
        result = scalekit.authenticate_with_code(
            code=code,
            redirect_uri='http://localhost:5000/auth/callback'
        )

        # Access user information
        id_token_claims = result.id_token_claims
        user = result.user

        # Step 3: Create session and store user info
        session['user_id'] = user.id
        session['email'] = user.email
        session['organization_id'] = id_token_claims.org_id

        return redirect('/dashboard')
    except Exception as error:
        print(f'Authentication failed: {error}')
        return redirect('/login?error=auth_failed')

# Step 4: Protect routes with authentication decorator
def require_auth(f):
    def decorated_function(*args, **kwargs):
        if 'user_id' not in session:
            return redirect('/login')
        return f(*args, **kwargs)
    decorated_function.__name__ = f.__name__
    return decorated_function

@app.route('/dashboard')
@require_auth
def dashboard():
    return f"Welcome, {session['email']}!"

# Step 5: Handle logout
@app.route('/auth/logout', methods=['POST'])
def logout():
    session.clear()
    return redirect('/login')

if __name__ == '__main__':
    app.run(port=5000)
